### PR TITLE
Require minimum variety of quest types [add AQuestDao test]

### LINKED ISSUE #1463: Require minimum variety of quest types
In some places after initial download user see quests of only one type.

It would be useful to ensure at least some variety. It is not obvious for people how many different quests are available and this protests from displaying just single type of quests.

Also, often there is a big cluster of some quests blocking further downloads due to high quest count but user is uninterested on solving them (at least at this specific moment).

I am opening an issue as it looks like more than finding relevant part of code and tweaking line or two, so I prefer to check is it going to be accepted.

So I proposed to add also min quest type count in a download strategy

To modify /app/src/main/java/de/westnordost/streetcomplete/data/download/QuestDownload.java (downloadQuestTypes)

Code doing now related thing: /app/src/main/java/de/westnordost/streetcomplete/data/osm/download/OsmQuestDownload.java

Place to write an SQL query: /app/src/main/java/de/westnordost/streetcomplete/data/osm/persist/OsmQuestDao.java

### PR DESCRIPTION
**Pull request by matkoniecz**
fixes #1463

I based it on `getCount` function. Possibly there is a better way to do that.

-----

My work on this pull request and UX testing was sponsored by a NGI Zero Discovery [grant](https://www.openstreetmap.org/user/Mateusz%20Konieczny/diary/368849)

### CODE REVIEW

File: app/src/main/java/de/westnordost/streetcomplete/data/download/MobileDataAutoDownloadStrategy.java [hunk 1]
```diff
@@ -37,4 +37,9 @@
 	{
 		return 600;
 	}
+
+	@Override
+	protected int getMinQuestsTypesInActiveRadius() {
+		return 2;
+	}
```
Comment (westnordost) [inline app/src/main/java/de/westnordost/streetcomplete/data/download/MobileDataAutoDownloadStrategy.java]: I think 2 will not make much of a difference to the current behavior. The auto downloader currently downloads 4 quest types until it checks again if there are now enough quests.

Question is, why is the current behavior not already enough to ensure a minimum variation in quest types? To find the answer to this question is important to know what should be changed to better the behavior.

Comment (matkoniecz) [reply app/src/main/java/de/westnordost/streetcomplete/data/download/MobileDataAutoDownloadStrategy.java]: > Question is, why is the current behavior not already enough to ensure a minimum variation in quest types? To find the answer to this question is important to know what should be changed in order to better the behavior.

I encountered case where multiple downloaded quests generated 0 results and road name quest generated big cluster of markers.

It was bad in two ways - (1) cluster was close geographically but not in place where someone wanted to walk (2) tester believed that app will continue to ask solely about street names.

Comment (westnordost) [reply app/src/main/java/de/westnordost/streetcomplete/data/download/MobileDataAutoDownloadStrategy.java]: ok

**Comment by matkoniecz** [comment]
> You tested it, right?

Yes, as usual I checked whatever new feature actually works (in this case I temporarily bumped number of required quests to 10 to skip looking for place where with specific data pattern)

**Comment by matkoniecz** [comment]
Now I see that while there are claims in log like `"Not downloading quests because there are enough quests in " + radius + "m radius"` it is actually checking in `enclosingBoundingBox(pos, radius)`.

I missed it earlier and noticed it now, but I have no good idea how to formulate it better.

**Comment by westnordost** [comment]
That's fine though

**Comment by westnordost** [comment]
I'll close this, because this is made obsolete by the Multi download

**Pull request closed without merging**